sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala

```diff
@@ -17,38 +17,84 @@
  */
 case class BatchScanExec(
     output: Seq[AttributeReference],
-    @transient scan: Scan) extends DataSourceV2ScanExecBase {
+    @transient scan: Scan,
+    dynamicFilters: Seq[Expression]) extends DataSourceV2ScanExecBase {
 
   @transient lazy val batch = scan.toBatch
 
+  private lazy val runnableDynamicFilters = dynamicFilters.collect {
+    case e: DynamicPruningExpression if e.child != Literal.TrueLiteral => e
+  }
+
   // TODO: unify the equal/hashCode implementation for all data source v2 query plans.
   override def equals(other: Any): Boolean = other match {
-    case other: BatchScanExec => this.batch == other.batch
+    case other: BatchScanExec =>
+      this.batch == other.batch && this.dynamicFilters == other.dynamicFilters
     case _ => false
   }
 
-  override def hashCode(): Int = batch.hashCode()
+  override def hashCode(): Int = Objects.hashCode(batch, dynamicFilters)
+
+  @transient private lazy val originalPartitions = batch.planInputPartitions()
+  @transient private var filteredPartitions: Option[Seq[InputPartition]] = None
 
-  @transient override lazy val partitions: Seq[InputPartition] = batch.planInputPartitions()
+  override def partitions: Seq[InputPartition] = filteredPartitions.getOrElse(originalPartitions)
 
-  override lazy val readerFactory: PartitionReaderFactory = batch.createReaderFactory()
+  private lazy val originalReaderFactory = batch.createReaderFactory()
+  private var filteredReaderFactory: Option[PartitionReaderFactory] = None
+
+  override def readerFactory: PartitionReaderFactory = {
+    filteredReaderFactory.getOrElse(originalReaderFactory)
+  }
 
   override lazy val inputRDD: RDD[InternalRow] = {
     new DataSourceRDD(sparkContext, partitions, readerFactory, supportsColumnar, customMetrics)
   }
 
   override def doCanonicalize(): BatchScanExec = {
-    this.copy(output = output.map(QueryPlan.normalizeExpressions(_, output)))
+    this.copy(
+      output = output.map(QueryPlan.normalizeExpressions(_, output)),
+      dynamicFilters = QueryPlan.normalizePredicates(
+        dynamicFilters.filterNot(_ == DynamicPruningExpression(Literal.TrueLiteral)),
+        output))
+  }
+
+  override protected def doPrepare(): Unit = {
+    if (runnableDynamicFilters.nonEmpty) {
+      // TODO: this triggers a broadcast which we don't need
+      waitForSubqueries()
+
+      val dataSourceFilters = runnableDynamicFilters.flatMap {
+        case DynamicPruningExpression(e) => DataSourceStrategy.translateDynamicFilter(e)
+        case _ => None
+      }
+
+      if (dataSourceFilters.isEmpty) {
+        logWarning("Skipping dynamic filtering, could not derive source filters")
+        return
+      }
+
+      val filterableScan = scan.asInstanceOf[SupportsDynamicFiltering]
+      filterableScan.filter(dataSourceFilters.toArray)
+
+      val filteredBatch = scan.toBatch
+
+      filteredPartitions = Some(filteredBatch.planInputPartitions())
```

Comment:
This means data source will need to plan input partitions twice - not sure if it could be expensive. Another idea is Spark provide the input partitions and ask data source to filter on top of them, like:
```
InputPartition[] filter(Filter[] filters, InputPartition[] partitions);
```
but I guess this may be too restrictive, so feel free to ignore.

Also may worth checking whether the original input partitions is used for making decisions before it's updated to `filteredPartitions` (and whether that would change the original decision). I can see that `supportsColumnar` is using it but not sure if there are other places.

Reply:
Hmm, I guess for scan implementation, if it can support this filtering, maybe it can cache original input partitions and the second planning is cheaper?

Reply:
@viirya is correct. Usually, there is no second planning. Instead, existing input partitions that has been already planned are filtered using dynamic filters.

Reply:
I see. In that case maybe it's useful to make this a bit more clear for data source implementors (I'm not sure if there's enough signal for them that `planInputPartitions` will be called twice).

Reply:
Added a note on caching state to `Scan`.